### PR TITLE
Create event ticket types in one call

### PR DESCRIPTION
**Pull request by Timo614**
### Closes Issues:
Closes: #1156 

### Description:

The associated issue details a problem where upon creating events we may fail partway through the list of ticket types being created leaving things in an odd state. To fix this we need to handle creating all of the ticket types in one call. We have API middleware that rolls back any transaction where an error occurs so we just need to create these all in one call to get that transaction behavior working.

### Examples:

With incorrect data:
```
bigneon=# select * from ticket_types where event_id = '80ced596-5e6b-4b23-a973-aff92dcd4773';
 id | event_id | name | description | status | start_date | end_date | increment | limit_per_person | created_at | updated_at | price_in_cents | cancelled_at | parent_id | rank | visibility 
----+----------+------+-------------+--------+------------+----------+-----------+------------------+------------+------------+----------------+--------------+-----------+------+------------
(0 rows)

curl --header "Content-Type: application/json" --request POST --data '{"ticket_types": [{"capacity": 1, "description": "", "end_date": "2019-11-14T05:00:00.000", "increment": 1, "limit_per_person": 10, "name": "Test 1", "parent_id": null, "price_in_cents": 2200, "start_date": "2022-04-29T16:00:00.000", "ticket_pricing": [], "visibility": "Always"}, {"capacity": 1, "description": "", "end_date": "2019-11-14T05:00:00.000", "increment": 1, "limit_per_person": 10, "name": "Test 2", "parent_id": null, "price_in_cents": 2200, "start_date": "2019-04-29T16:00:00.000", "ticket_pricing": [], "visibility": "Always"}]}' http://localhost:9000/events/80ced596-5e6b-4b23-a973-aff92dcd4773/ticket_types/multiple -i --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI5MDIwZjVhMi0zZjQyLTQ3ZTUtODYyMy03Yjg0YmNiNGE0NWYiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTU2NTY5Mjg5fQ.VSfhVRmZRX34CaAaL88jawfplMOscneXlEBJ-130gHY"
HTTP/1.1 422 Unprocessable Entity
content-length: 226
content-type: application/json
access-control-allow-origin: http://localhost:3000
access-control-expose-headers: x-app-version
vary: Origin
x-app-version: 1.0.89
date: Mon, 29 Apr 2019 20:06:46 GMT

{"error":"Validation error","fields":{"start_date":[{"code":"start_date_must_be_before_end_date","message":"Start date must be before end date","params":{"end_date":"2019-11-14T05:00:00","start_date":"2022-04-29T16:00:00"}}]}}

bigneon=# select * from ticket_types where event_id = '80ced596-5e6b-4b23-a973-aff92dcd4773';
 id | event_id | name | description | status | start_date | end_date | increment | limit_per_person | created_at | updated_at | price_in_cents | cancelled_at | parent_id | rank | visibility 
----+----------+------+-------------+--------+------------+----------+-----------+------------------+------------+------------+----------------+--------------+-----------+------+------------
(0 rows)
```

With correct data:
```
curl --header "Content-Type: application/json" --request POST --data '{"ticket_types": [{"capacity": 1, "description": "", "end_date": "2019-11-14T05:00:00.000", "increment": 1, "limit_per_person": 10, "name": "Test 1", "parent_id": null, "price_in_cents": 2200, "start_date": "2019-04-29T16:00:00.000", "ticket_pricing": [], "visibility": "Always"}, {"capacity": 1, "description": "", "end_date": "2019-11-14T05:00:00.000", "increment": 1, "limit_per_person": 10, "name": "Test 2", "parent_id": null, "price_in_cents": 2200, "start_date": "2019-04-29T16:00:00.000", "ticket_pricing": [], "visibility": "Always"}]}' http://localhost:9000/events/80ced596-5e6b-4b23-a973-aff92dcd4773/ticket_types/multiple -i --header "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI5MDIwZjVhMi0zZjQyLTQ3ZTUtODYyMy03Yjg0YmNiNGE0NWYiLCJpc3MiOiJiZy1zYW1wbGUtaXNzdWVyIiwiZXhwIjoxNTU2NTY5Mjg5fQ.VSfhVRmZRX34CaAaL88jawfplMOscneXlEBJ-130gHY"
HTTP/1.1 201 Created
content-length: 93
content-type: application/json
access-control-allow-origin: http://localhost:3000
access-control-expose-headers: x-app-version
vary: Origin
x-app-version: 1.0.89
date: Mon, 29 Apr 2019 20:07:50 GMT

[{"id":"7cfcf30d-b9c3-428b-91d8-b9d5526e90de"},{"id":"17b8361f-3799-4113-b4c1-435e8765b292"}]

bigneon=# select * from ticket_types where event_id = '80ced596-5e6b-4b23-a973-aff92dcd4773';
                  id                  |               event_id               |  name  | description |  status   |     start_date      |      end_date       | increment | limit_per_person |         created_at         |         updated_at         | price_in_cents | cancelled_at | parent_id | rank | visibility 
--------------------------------------+--------------------------------------+--------+-------------+-----------+---------------------+---------------------+-----------+------------------+----------------------------+----------------------------+----------------+--------------+-----------+------+------------
 7cfcf30d-b9c3-428b-91d8-b9d5526e90de | 80ced596-5e6b-4b23-a973-aff92dcd4773 | Test 1 |             | Published | 2019-04-29 16:00:00 | 2019-11-14 05:00:00 |         1 |               10 | 2019-04-29 20:07:50.377273 | 2019-04-29 20:07:50.377273 |           2200 |              |           |    1 | Always
 17b8361f-3799-4113-b4c1-435e8765b292 | 80ced596-5e6b-4b23-a973-aff92dcd4773 | Test 2 |             | Published | 2019-04-29 16:00:00 | 2019-11-14 05:00:00 |         1 |               10 | 2019-04-29 20:07:50.377273 | 2019-04-29 20:07:50.377273 |           2200 |              |           |    2 | Always
(2 rows)
```

### Migrations
 * No Migrations

### Environment Variables
 * No change

### CODE REVIEW

File: api/src/controllers/ticket_types.rs [hunk 1]
```diff
@@ -153,6 +158,95 @@ pub fn create(
     Ok(HttpResponse::Created().json(DisplayCreatedTicket { id: ticket_type.id }))
 }
 
+pub fn create_multiple(
```
Comment (Krakaw) [inline api/src/controllers/ticket_types.rs]: From a DRY point of view could we not keep this code in here as is, but then from the [create](https://github.com/big-neon/bn-api/blob/b1ea934a1432a23ba9438369bf0ab0e6b7769f3c/api/src/controllers/ticket_types.rs#L81) function call
```rust
let data = json!(CreateMultipleTicketTypeRequest {ticket_types: vec![data]});
create_multiple((connection, path, data, user, state));
...

```

Comment (Timo614) [reply api/src/controllers/ticket_types.rs]: We'd end up doing the scope check call once per ticket type created which wouldn't be ideal (although SQL caching would prevent some of that).

There isn't too much going on with the controller just some very large method signatures due to us translating the structs over to create methods.

I'll DRY up some of the logic by moving some of the creation logic into methods that can be shared between the two though which should be helpful here.

Comment (Timo614) [reply api/src/controllers/ticket_types.rs]: Actually I see what you mean, so having the single one call the multiple as well. Yeah that would be fine, will switch that.

Comment (Timo614) [reply api/src/controllers/ticket_types.rs]: I DRYed it up a bit by having both methods call a method that creates the ticket types. Opted not to just call the multiple controller action as it returns a HttpResponse so I'd have had to parse out the ticket type id via serde for the response. This just felt a bit cleaner. I kept the logic in place per controller to check the permissions as it feels more complete that way given putting it in the shared method would obscure it further.

Comment (Krakaw) [reply api/src/controllers/ticket_types.rs]: I like that the controllers maintain their permission checks, it makes it clearer to someone viewing the code that there are limitations on that function. Does it make any sense to put that shared logic in the model?

**Pull request merged**